File: populating-next-right-pointers-in-each-node.js
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
    if (!root) return root;
    
    const q = [root];
    while (q.length > 0) {
        // console.log(q);
        let level = q.length;
        for (let i = 0; i < level; i++) {
            let curr = q.shift();
            if (i < level - 1) curr.next = q[0];
            else curr.next = null;
            
            if (curr.left) q.push(curr.left, curr.right);
        }
    } 
    return root;
};

File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
​
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        
        def dfs(left,right):
            #base
            if left == None:
                return
            
            #logic
            left.next = right
            dfs(left.left,left.right)
            dfs(left.right,right.left)
            dfs(right.left,right.right)
            
        
        if root is None:
            return root
        
        dfs(root.left,root.right)
        return root
        
